Abstract restriction logic in a new Restricter class

Murderlon opened the pull request:
The Uppy class has become too big and hard to navigate and does too many things for one class. It handles events, restrictions, uploading (with pre and post processors, and pause/resume), events, state management, and methods for convenience (getting state into a certain data structure).

To improve this without breaking changes and introducing a faulty inheritance model, I chose to abstract grouped logic (logic that changes together) into a class and apply dependency injection to share variables/functions.

- Create `Restriction` class and move all related methods and data into there. Pass `getOpts` and `i18n` in the constructor. 
- Create a private instance of the restriction class in Uppy.
- Private restriction related methods from the Uppy class can be removed entirely
- Restrictor logic is now delegated but side effects still happen in core

### Future improvements

- Do the same for uploading
- Do the same for events

aduh95 commented on packages/@uppy/core/src/Restricter.js:
```diff
@@ -0,0 +1,143 @@
+}
+
+class AggregateRestrictionError extends AggregateError {
+  constructor (...args) {
+    super(...args)
+    this.isRestriction = true
+  }
```
Unrelated, but should we do this instead?
```suggestion
  isRestriction = true
```

aduh95 commented on packages/@uppy/core/src/Restricter.js:
```diff
@@ -0,0 +1,143 @@
+}
+
+class RestrictionError extends Error {
+  constructor (...args) {
+    super(...args)
+    this.isRestriction = true
+  }
```
```suggestion
  isRestriction = true
```

aduh95 commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -665,7 +506,11 @@ class Uppy {
+      this.#restricter.validate(newFile, filesArray, {
+        opts: this.opts,
+        i18n: this.i18n,
+        getFiles: this.getFiles.bind(this),
```
This is never used?
```suggestion
```

aduh95 reviewed (commented):
We should get rid of the bind calls.

mifi commented on packages/@uppy/core/src/Restricter.js:
```diff
@@ -0,0 +1,143 @@
+
+if (typeof AggregateError === 'undefined') {
+  // eslint-disable-next-line no-global-assign
+  globalThis.AggregateError = class AggregateError extends Error {
```
would it be cleaner to instead export this error and let core assign it to the global object? Actually assigning it to the global object is kinda dirty but I guess it's part of our public API so it would be a breaking change if we instead export it from core?

mifi replied:
Should we instead recommend that users polyfill this class? Or use the core.js polyfill instead of reimplementing it ourselves?

aduh95 replied:
What would be the upside? There's probably way more overhead to add a core.js polyfill rather than this 4 liner one.

mifi replied:
I can see that the implementation in core.js seems more complex than ours: https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.aggregate-error.js

maybe our simple implementation is not spec-complete? Then our code could behave differently depending on whether the browser supports it or not

aduh95 replied:
OK agreed. Let's add a TODO to remove this in the next major, folks who want to support browsers that have not implemented it can add the polyfill themselves, to align with our policy on polyfills in Uppy.

aduh95 replied:
```suggestion
  // TODO: remove this "polyfill" in the next major.
  globalThis.AggregateError = class AggregateError extends Error {
```

mifi commented on packages/@uppy/core/src/Restricter.js:
```diff
@@ -0,0 +1,143 @@
+  }
+}
+
+class Restricter {
```
would it better to just export the functions instead of wrapping them in a class? I don't see the benefit of using a class here. then we could also remove `eslint-disable class-methods-use-this `

Murderlon replied:
I'm inclined to keep it a class to do this: 
> There are repetitive things being injected, such as `i18n` and `opts`. Perhaps better to add these in the constructor.

mifi replied:
Alright, with contructor arguments it makes sense to use a class. However this class now has two ways of passing dependencies: one is through the constructor, and the other is through `ctx`. Should we pass the ctx properties also in through the constructor to make it consistent?

mifi commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -7,41 +7,22 @@ const Translator = require('@uppy/utils/lib/Translator')
 const { justErrorsLogger, debugLogger } = require('./loggers')
+const {
+  Restricter,
+  defaultOptions: defaultRestrictionOptions,
```
could the default options instead be declared per method inside the restricter module? that way core wouldn't need to have any knowledge about that object.

```js
checkRequiredMetaFieldsOnFile (file, requiredMetaFields = []) {
   ...
}
```

Murderlon replied:
I don't think this would work. The properties need to exist because other plugins, like dashboard, expect it in `opts`.

Murderlon commented:
- Removed the `bind` calls with arrow functions. Although I don't like it as much `bind`. It's more verbose and implicit. With `bind` at least you're telling 'hey we are switching context' and an arrow function is just an arrow function
- Moved `opts` and `i18n` into the constructor. This made half the functions not needing `ctx` at all. However, one should only update `opts` through `setOptions`, any other way would not be detected as I update `restricter.opts` only there.
- Tiny cleanups

mifi commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -261,6 +232,8 @@ class Uppy {
 
     this.i18nInit()
 
+    this.#restricter.opts = this.opts
```
Instead of reassigning this all the time, maybe in the Restrictors constructor we could instead pass a function `getOptions: () => this.opts` that it can call. that way we don't have to duplicate the state in two places?

aduh95 commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -1242,7 +1080,11 @@ class Uppy {
       if (file) {
-        this.#checkRequiredMetaFieldsOnFile(file)
+        const ctx = {
+          setFileState: (...args) => this.setFileState(...args),
```
Alternatively, we could define a private property that to make it clear it's a bound method
```suggestion
          setFileState: this.#setFileStateBound,
```

And add below the `setFileState` definition:

```js
  #setFileStateBound = (...args) => this.setFileState(...args)
```

Murderlon commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -583,6 +411,31 @@ class Uppy {
+    }
+  }
+
+  checkRequiredMetaFields (files) {
+    Object.values(files).forEach((file) => this.checkRequiredMetaFieldsOnFile(file))
+  }
```
Note that I simplified this, it no longer has this: 

```js

    if (errors.length) {
      throw new AggregateRestrictionError(errors, `${this.i18n('missingRequiredMetaField')}`)
    }
```

I don't understand the purpose of `AggregateRestrictionError`, we have no tests for it, and it's never used in the codebase.

aduh95 replied:
The `AggregateError` allows to display one error instead of a dozens of errors. I think we should keep it, or at least provide a way not to overwhelm users if there are several errors with their files.

Murderlon replied:
I'm not sure. Currently things are being logged in `checkRequiredMetaFieldsOnFile` (which is called by `checkRequiredMetaFields`):

```js
    errorMap.forEach((_, error) => {
      this.showOrLogErrorAndThrow(error, { file, showInformer: false, throwErr: false })
    })
```

So all errors are already reported. If you throw an error it will be picked up and then all errors will be reported _again_. I think if the underlying function doesn't throw, then the top one shouldn't either, if it's merely a wrapper to iterate over files.

aduh95 replied:
As you can see `showInformer` is `false`, and `throwErr` is `false`, so I don't think that error would be reported with that call. IIRC it would save those errors in an aggregator and only report the aggregated error (unless there was one one error).

Murderlon replied:
The problem was, when looking deeper, we never aggregated those errors. Nor did it have the desired effect (obscured by `showOrLogErrorAndThrow`). I now emit all the files that failed as a normal restriction error. I will write some docs to clarify this. (TypeScript overloading would be great for this)

aduh95 commented on packages/@uppy/core/src/Restricter.js:
```diff
@@ -0,0 +1,127 @@
+    if (allowedFileTypes) {
+      const isCorrectFileType = allowedFileTypes.some((type) => {
+        // check if this is a mime-type
+        if (type.indexOf('/') > -1) {
```
```suggestion
        if (type.includes('/')) {
```

aduh95 commented on packages/@uppy/core/src/Restricter.js:
```diff
@@ -0,0 +1,127 @@
+
+  validateFile (file) {
+    const { requiredMetaFields } = this.getOpts().restrictions
+    const own = Object.prototype.hasOwnProperty
```
```suggestion
    // TODO: migrate to Object.hasOwn in the next major.
    const own = Object.prototype.hasOwnProperty
```

aduh95 commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -154,6 +123,9 @@ class Uppy {
       recoveredState: null,
     })
 
+    const getOpts = () => this.opts
+    this.#restricter = new Restricter(getOpts.bind(this), this.i18n)
```
```suggestion
    this.#restricter = new Restricter(() => this.opts, this.i18n)
```
but wouldn't it be easier to do?
```suggestion
    this.#restricter = new Restricter(this)
```

Murderlon replied:
I don't like this, it should be explicit. I think what I did is fine.

aduh95 replied:
Well I think my first suggestion is better and still explicit 😅 Did you miss it?
```suggestion
    this.#restricter = new Restricter(() => this.opts, this.i18n)
```

Murderlon replied:
I tried that and it and it doesn't work.

Murderlon commented:
I have to think to get rid of some final error handling inconsistencies here (without creating breaking changes)

mifi commented on packages/@uppy/core/src/Restricter.js:
```diff
@@ -0,0 +1,129 @@
+  }
+}
+
+class Restricter {
```
now that all the methods are named "validateX" should we also rename this to Validator? :)

Murderlon replied:
I think Restricter is conceptually okay because it exclusively deals with `opts.restrictions` and it throws errors, restricting an upload.

mifi reviewed (commented):
starting to look good now :) left a few comments

Murderlon commented:
Note that I got rid of `showOrLogErrorAndThrow`, it did too many things and obscured what would happen with error handling. Instead I made it more explicit at every usage. Going through all of this showed to me that we are very inconsistent with error handling but it is out of scope for this PR. I will gradually improve it, next up will be uploading which as most of the inconsistencies. I also got rid of `AggregateError` as we never correctly implemented this. Instead, I throw a `RestrictionError` which will send all the files along with the `restriction-error` event.

aduh95 commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -400,194 +371,73 @@ class Uppy {
+    try {
+      this.#restricter.validate(file, files)
+      return { result: true }
+    } catch (err) {
+      return { result: false, reason: err.message }
```
Shouldn't we send the whole error object? What if the `err` is nullish?
```suggestion
    } catch (reason) {
      return { result: false, reason }
```

Murderlon replied:
Yes I find this very weird, but like I said in the comment above changing this is a breaking change. Definitely a todo for the next major

aduh95 commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -400,194 +371,73 @@ class Uppy {
+    try {
+      this.#restricter.validate(file, files)
+      return { result: true }
+    } catch (err) {
+      return { result: false, reason: err.message }
```
Shouldn't we send the whole error object? What if the `err` is nullish?
```suggestion
    } catch (reason) {
      return { result: false, reason }
```

aduh95 commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -400,194 +371,73 @@ class Uppy {
+  * @param { Array<UppyFile> } fileOrFiles
+  */
+  #informAndEmit (error, fileOrFiles) {
+    const { message, details = '' } = error
```
We should first check that `error` is not nullish.

mifi replied:
if error is nullish, wouldn't that be a bug, so crashing could be nice?

Murderlon replied:
This was one of the reasons I refactored `showOrLogAndThrow`, it could be any kind of error, string, undefined, whatever. I know changed this to always requiring an actual error.

mifi replied:
This change might be breaking something, but I guess we can take this risk in order to expose any unknown bugs (where string, undefined, null etc is thrown somewhere)

Murderlon replied:
It won't be breaking because the method is private and I refactored all instances myself where this is called to make sure it passes an actual error.

aduh95 commented on packages/@uppy/core/src/Uppy.js:
```diff
@@ -400,194 +371,73 @@ class Uppy {
+  * @param { Array<UppyFile> } fileOrFiles
+  */
+  #informAndEmit (error, fileOrFiles) {
+    const { message, details = '' } = error
```
We should first check that `error` is not nullish.

aduh95 reviewed (commented):
A few comments to improve code that was not necessarily added by this PR, but since this is a refactor any way, let's refactor this as well.

mifi commented on packages/@uppy/core/src/Restricter.js:
```diff
@@ -0,0 +1,132 @@
+  validate (file, files) {
+    const { maxFileSize, minFileSize, maxTotalFileSize, maxNumberOfFiles, allowedFileTypes } = this.getOpts().restrictions
+
+    if (maxNumberOfFiles) {
+      if (files.length + 1 > maxNumberOfFiles) {
```
I think eslint is supposed to deal with these kind of issues 🤔 maybe we disabled that rule
